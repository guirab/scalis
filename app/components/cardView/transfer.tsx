"use client";
import { useState } from "react";

import { TransferSameAcc } from "../transfer/sameAcc";
import { TransferOtherAcc } from "../transfer/otherAcc";

export const TransferView: React.FC<React.HTMLProps<any> & ActionCardType> = ({
  open,
  setOpen,
  ...props
}) => {
  const [isSameAccount, setIsSameAccount] = useState("true");

  const currentView = {
    true: <TransferSameAcc open={open} setOpen={setOpen} />,
    false: <TransferOtherAcc open={open} setOpen={setOpen} />,
  }[isSameAccount];

  return (
    <>
      {open && (
        <div className="block w-full mt-4 space-y-4" {...props}>
          <div className="gap-x-2 mb-4 flex text-sm justify-between">
            <div className="flex justify-start items-center">
              <label className="pr-px" htmlFor="same">
                This account
              </label>
              <input
                data-testid="same-account"
                type="radio"
                id="same"
                value="same"
                checked={isSameAccount === "true"}
                onChange={() => setIsSameAccount("true")}
              />
            </div>
            <div className="flex justify-start items-center">
              <label className="pr-px" htmlFor="other">
                Other account
              </label>
              <input
                data-testid="other-account"
                type="radio"
                id="other"
                value="other"
                checked={isSameAccount === "false"}
                onChange={() => setIsSameAccount("false")}
              />
            </div>
          </div>
          {currentView}
        </div>
      )}
    </>
  );
};
