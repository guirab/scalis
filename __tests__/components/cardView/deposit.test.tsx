import {
  render,
  screen,
  fireEvent,
  cleanup,
  waitFor,
} from "@testing-library/react";
import { DepositView } from "../../../app/components/cardView/deposit";
import { afterEach, describe, expect, test, vitest } from "vitest";

describe("DepositView", () => {
  afterEach(() => {
    cleanup();
  });
  test("snapshot", () => {
    const { container } = render(<DepositView open setOpen={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  test("updates the type state when the select value changes", () => {
    render(<DepositView open setOpen={() => {}} />);

    const select = screen.getByTestId("deposit-select") as HTMLSelectElement;

    fireEvent.change(select, {
      target: { value: "savings" },
    });
    expect(select.value).toBe("savings");

    fireEvent.change(select, {
      target: { value: "checking" },
    });
    expect(select.value).toBe("checking");
  });

  test("updates the amount state when the input value changes", () => {
    render(<DepositView open setOpen={() => {}} />);

    const input = screen.getByLabelText("Amount:") as HTMLInputElement;

    fireEvent.change(input, { target: { value: "100" } });
    expect(input.value).toBe("1.00");
  });

  test("calls the onClick function when the button is clicked", async () => {
    vitest.mock("react", async () => {
      return {
        ...(await vitest.importActual("react")),
        useContext: () => ({
          accounts: [{ id: 1, username: "test" }],
          setAccounts: vitest.fn(),
        }),
      };
    });

    const setOpen = vitest.fn();
    render(<DepositView open setOpen={setOpen} />);

    const input = screen.getByLabelText("Amount:") as HTMLInputElement;

    const transferToSameAcc = vitest.fn().mockResolvedValue({});

    fireEvent.change(input, {
      target: { value: "100" },
    });

    fireEvent.click(screen.getByRole("button", { name: "Deposit" }));

    await waitFor(() => {
      expect(transferToSameAcc).toHaveBeenCalledWith({
        action: "deposit",
        type: "checking",
        amount: 100,
        id: 1,
      });
    });

    expect(setOpen).toHaveBeenCalledWith(false);
  });
});
