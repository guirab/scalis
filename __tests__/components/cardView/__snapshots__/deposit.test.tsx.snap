// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DepositView > snapshot 1`] = `
<div>
  <div>
    <label
      for="deposit"
    >
      Into:  
    </label>
    <select
      class="text-black pl-2 outline-none rounded-md w-full"
      data-testid="deposit-select"
      id="deposit"
      name="deposit"
    >
      <option
        value="checking"
      >
        Checking
      </option>
      <option
        value="savings"
      >
        Savings
      </option>
    </select>
    <div>
      <label
        for="currency-deposit"
      >
        Amount:  
      </label>
      <div
        class="relative w-full"
      >
        <span
          class="absolute text-black left-px"
        >
          $ 
        </span>
        <input
          autocomplete="off"
          class="text-black pl-4 text-black pl-2 appearance-none outline-none rounded-md w-full"
          data-testid="currency-input"
          id="currency-deposit"
          name="currency-deposit"
          required=""
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      class="bg-blue-500 text-white px-4 py-2 rounded-md mt-4 w-full"
    >
      Deposit
    </button>
  </div>
</div>
`;
